### ISSUE TITLE
Crash with generator in try finally

### ISSUE THREAD

**Issue by Resery** [Nuitka/Nuitka#1239]
Nuitka version
0.6.17rc3
Python: 3.7.3 (default, Sep  8 2021, 19:25:40) 
OS: Linux
Arch: x86_64

Nuitka Options
python -m nuitka --clang test.py

```
dev@dev:~/$ cat test.py
def returnintry():
	try:
		print("test")
		return 9
	finally:
		print(''.join([chr(var1 ^ var2) for var2, var1 in zip(b"ba", b"ab")]))
		print(''.join(chr(var1 ^ var2) for var2, var1 in zip(b"ba", b"ab")))

returnintry()
dev@dev:~/$ python3 test.py
test


dev@dev:~/$ ./test.bin
test

Illegal instruction (core dumped)
```

**Comment by Resery** [comment]
> @Resery why did you close the issue?

Because the issue has not been answered for a long time so I think my question may be worthless

**Comment by kayhayen** [comment]
It's not.

**Comment by kayhayen** [comment]
I managed to drill it down to this:

```
def returnintry():
    try:
        return 9
    finally:
        print("".join(chr(var1 ^ var2) for var2, var1 in zip(b"ba", b"ab")))


returnintry()
```

And what's happening is that this code gets generated, about which the C compiler also warns for use of uninitialized value:

```
        // Tried code:
        {
            struct Nuitka_CellObject *tmp_closure_1[1];
            Py_DECREF(tmp_args_element_name_2);

            tmp_closure_1[0] = Nuitka_Cell_New0(tmp_genexpr_1__$0);

            tmp_args_element_name_2 = MAKE_GENERATOR___main__$$$function__1_returnintry$$$genexpr__1_genexpr(tmp_closure_1);

            goto try_return_handler_2;
        }
```

That seems to include, that Nuitka believes there is an old value that needs releasing, and that's of course untree. I am pretty much shocked, there is no loop in there, this must be a very severe issue.

**Comment by kayhayen** [comment]
I believe, I found a strangeness that points to the problem. There is code generated, following `try` ends, which is unreachable, and most likely confusing the state of code generator to believe a variable was assigned, and needs release:

```
        NUITKA_PRINT_TRACE("Execute: " "MiniTryGeneratorCrash.py:5 <Node 'STATEMENT_RETURN_RETURNED_VALUE'>");
        goto outline_result_2;
        // End of try:
        NUITKA_PRINT_TRACE("Execute: " "MiniTryGeneratorCrash.py:5 <Node 'STATEMENT_RELEASE_VARIABLE' with {'variable_name': 'genexpr_1__.0', 'owner': '__main__$$$function__1_returnintry'}>");
        CHECK_OBJECT(tmp_genexpr_1__$0);
        Py_DECREF(tmp_genexpr_1__$0);
        tmp_genexpr_1__$0 = NULL;
        NUITKA_CANNOT_GET_HERE("Return statement must have exited already.");
```

This seems to actually happen relatively many times. The "outline" is the creating of a generator object from the iterated argument, and the release that's made there, makes no sense, and was generated by that outline body. This might be about not removing dead code, although said try gets code generated at the end of its try handler:

```
        NUITKA_CANNOT_GET_HERE("tried codes exits in all cases");
        return NULL;
```

So, need to check where that release of the argument is coming from.

**Comment by kayhayen** [comment]
I believe, I found a strangeness that points to the problem. There is code generated, following `try` ends, which is unreachable, and most likely confusing the state of code generator to believe a variable was assigned, and needs release:

```
        NUITKA_PRINT_TRACE("Execute: " "MiniTryGeneratorCrash.py:5 <Node 'STATEMENT_RETURN_RETURNED_VALUE'>");
        goto outline_result_2;
        // End of try:
        NUITKA_PRINT_TRACE("Execute: " "MiniTryGeneratorCrash.py:5 <Node 'STATEMENT_RELEASE_VARIABLE' with {'variable_name': 'genexpr_1__.0', 'owner': '__main__$$$function__1_returnintry'}>");
        CHECK_OBJECT(tmp_genexpr_1__$0);
        Py_DECREF(tmp_genexpr_1__$0);
        tmp_genexpr_1__$0 = NULL;
        NUITKA_CANNOT_GET_HERE("Return statement must have exited already.");
```

This seems to actually happen relatively many times. The "outline" is the creating of a generator object from the iterated argument, and the release that's made there, makes no sense, and was generated by that outline body. This might be about not removing dead code, although said try gets code generated at the end of its try handler:

```
        NUITKA_CANNOT_GET_HERE("tried codes exits in all cases");
        return NULL;
```

So, need to check where that release of the argument is coming from.

**Comment by kayhayen** [comment]
Ok, this is just something that is wrong about its checking if it's aborting. The absence of an exception handler is treated as if it was absorbing the exception (no statement, no re-raise), but of course `return 9` is never going to raise an exception, so that assessment was wrong. I made an improvement to `isStatementAborting()` and `mayReturn`, `mayContinue` that also involves caching that decision, such that if it was once found to not raise, we do not ask again, as it won't start to raise again. That might also improve compile time performance.

However, this doesn't really solve the bug, but it's more clearly visible now.

**Comment by kayhayen** [comment]
Something called "return release mode" decides to release the return value before its being assigned. There must be some ordering problem with this mode and return statements, the generate expression is an inlined function that also did a return. From my inspection it properly sets up and restores the mode, but there must be some bug related to that. I think I am getting close finally.

**Comment by kayhayen** [comment]
Ok, I found it, when an outline is code generated, Nuitka restored the return release mode as it existing on the outside, but it also needs to restart freshly for the outline return, and a "true" that doesn't impact its return value name correctly would leak inside the finally of tried blocks that return.

The fix is on factory and will be released as part of a hotfix for 0.6.17 as soon as my CI liked factory:

https://nuitka.net/doc/factory.html

**Comment by kayhayen** [comment]
Part of the 0.6.17.7 hotfix.
